jdbc/src/main/java/org/apache/metamodel/jdbc/dialects/HiveQueryRewriter.java

```diff
@@ -30,6 +35,59 @@ public HiveQueryRewriter(JdbcDataContext dataContext) {
         super(dataContext);
     }
 
+    @Override
+    public String rewriteQuery(Query query) {
+
+        Integer maxRows = query.getMaxRows();
+        Integer firstRow = query.getFirstRow();
+
+        if(firstRow != null && firstRow > 1) {
+            if(query.getOrderByClause().getItemCount() == 0) {
+                throw new MetaModelException("OFFSET requires an ORDER BY clause");
+            }
+        }
+
+        if (maxRows == null && (firstRow == null || firstRow.intValue() == 1)) {
+            return super.rewriteQuery(query);
+        }
+
+        if ((firstRow == null || firstRow.intValue() == 1) && maxRows != null && maxRows > 0) {
+            // We prefer to use the "LIMIT n" approach, if
+            // firstRow is not specified.
+            return super.rewriteQuery(query) + " LIMIT " + maxRows;
+        }
+
+        final Query innerQuery = query.clone();
+        innerQuery.setFirstRow(null);
+        innerQuery.setMaxRows(null);
+
+        final Query outerQuery = new Query();
+        final FromItem subQuerySelectItem = new FromItem(innerQuery).setAlias("metamodel_subquery");
+        outerQuery.from(subQuerySelectItem);
+
+        final List<SelectItem> innerSelectItems = innerQuery.getSelectClause().getItems();
+        for (SelectItem selectItem : innerSelectItems) {
+            outerQuery.select(new SelectItem(selectItem, subQuerySelectItem));
+        }
+
+
+        final String rewrittenOrderByClause = rewriteOrderByClause(innerQuery, innerQuery.getOrderByClause());
+        final String rowOver = "ROW_NUMBER() " + "OVER(" + rewrittenOrderByClause + ")";
+        innerQuery.select(new SelectItem(rowOver, "metamodel_row_number"));
+        innerQuery.getOrderByClause().removeItems();
+
+        final String baseQueryString = rewriteQuery(outerQuery);
+
+        if (maxRows == null) {
+            return baseQueryString + " WHERE metamodel_row_number > " + (firstRow - 1);
+        }
+
+        return baseQueryString + " WHERE metamodel_row_number BETWEEN " + firstRow + " AND "
```

Comment:
I notice now how most of this code is duplicated with `DB2QueryRewriter`. Please make sure that the code is shared via a common super-class or a utility function or something like that.